File: Structural/Decorator/Swift/WindowDecorator.swift
import Foundation

class WindowDecorator: Window
{
  var targetWindow:IWindow

  init(targetWindow:IWindow, title:String)
  {
    self.targetWindow = targetWindow
    super.init(title: title)
  }
}

class ScrollDecorator: WindowDecorator
{
  override init(targetWindow: IWindow, title: String) {
    super.init(targetWindow: targetWindow, title: title)
  }

  func Render()
  {
    self.targetWindow.Render()
  }

  func ScrollBy(amaount: Int)
  {
    print("\(targetWindow.Title) penceresi \(amaount) birim kaydırıldı")
    Render()
  }
}

class ThemeDecorator: WindowDecorator
{
  override init(targetWindow: IWindow, title: String) {
    super.init(targetWindow: targetWindow, title: title)
  }

  func Render()
  {
    self.targetWindow.Render()
  }

  func SetTheme(name: String)
  {
    print("\(targetWindow.Title) başlıklı pencereye \(name) teması uygulandı")
    Render()
  }
}

protocol IWindow
{
    var Title:String
    {
        get
        set
    }

    func Render()
}

class Window
{
    fileprivate var title:String

    var Title:String
    {
        get
        {
            return title
        }
        set (value)
        {
            title = value
        }
    }

    init(title:String) {
        self.title = title
    }

}

class LoginWindow: Window, IWindow
{
    override init(title:String) {
        super.init(title:title)
    }

    func Render ()
    {
        print(title + "login ekranı render edildi")
    }

}

class ReportWindow: Window, IWindow
{
    override init(title:String) {
        super.init(title:title)
    }

    func Render ()
    {
        print(title + "rapor ekranı render edildi")
    }

}


// TEST
var wnd1:IWindow = LoginWindow(title: "Sisteme Giriş")
var wnd2:IWindow = ReportWindow(title: "Stok Raporu Ekranı")
var wnd3:IWindow = ReportWindow(title: "Satış Raporu Ekranı")

var decorator1:ScrollDecorator = ScrollDecorator(targetWindow: wnd2, title: "Stok Raporu Ekranı")

var decorator2:ThemeDecorator = ThemeDecorator(targetWindow: wnd1, title: "Sisteme Giriş")

var decorator3:ThemeDecorator = ThemeDecorator(targetWindow: wnd3, title: "Satış Raporu Ekranı")

var decorator4:ThemeDecorator = ThemeDecorator(targetWindow: wnd2, title: "Stok Raporu Ekranı")

decorator1.ScrollBy(amaount: 5)
decorator2.SetTheme(name: "Gökyüzü")
decorator3.SetTheme(name: "Sonbahar")
decorator4.SetTheme(name: "Deniz")

File: Structural/Decorator/Kotlin/WindowDecorator.kt
abstract class WindowDecorator(targetWindow: Window,
                               title:String): Window()
{
    var targetWindow: Window
    init{
        this.targetWindow = targetWindow
        this.targetWindow.title = title
    }
}

class ScrollDecorator(window: Window,
                      title:String): WindowDecorator(window, title)
{
    public override fun Render() {
        this.targetWindow.Render()
    }

    fun ScrollBy(amount: Int) {
        println(String.format("%s başlıklı pencere %d birim kaydırıldı",
            targetWindow.title, amount))
        Render()
    }
}

class ThemeDecorator(window: Window,
                     title:String): WindowDecorator(window, title)
{
    public override fun Render() {
        this.targetWindow.Render()
    }

    fun SetTheme(name: String) {
        println(String.format("%s başlıklı pencereye %s teması uygulandı",
            targetWindow.title, name))
        Render()
    }
}

abstract class Window
{
    var title: String? = ""

    constructor(): super() {}

    constructor(title: String) {
        this.title = title
    }

    abstract fun Render()
}

class LoginWindow: Window
{
    constructor() {}

    constructor(title: String): super(title) {}

    public override fun Render() {
        println(title + "başlıklı login ekranı render edildi")
    }
}

class ReportWindow: Window
{
    constructor() {}

    constructor(title: String): super(title) {}

    public override fun Render() {
        println(title + "başlıklı rapor ekranı render edildi")
    }
}


// test
object Program
{
    @JvmStatic fun main(args:Array<String>)
    {
        val wnd1 = LoginWindow("Sisteme Giriş")
        val wnd2 = ReportWindow("Stok Raporu Ekranı")
        val wnd3 = ReportWindow("Satış Raporu Ekranı")

        val decorator1 = ScrollDecorator(wnd2, "Stok Raporu Ekranı")

        val decorator2 = ThemeDecorator(wnd1, "Sisteme Giriş")

        val decorator3 = ThemeDecorator(wnd3, "Satış Raporu Ekranı")

        val decorator4 = ThemeDecorator(wnd2, "Stok Raporu Ekranı")

        decorator1.ScrollBy(5)
        decorator2.SetTheme("Gökyüzü")
        decorator3.SetTheme("Sonbahar")
        decorator4.SetTheme("Deniz")
    }
}